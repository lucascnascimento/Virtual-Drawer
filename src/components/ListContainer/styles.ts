import styled from 'styled-components/native';

export const Separator = styled.View`
  flex: 1;
  max-height: 1px;
  background-color: #e4e4e4;
`;
