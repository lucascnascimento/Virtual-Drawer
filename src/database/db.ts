import { Alert } from 'react-native';
import {
  openDatabase,
  ResultSet,
  SQLiteDatabase,
} from 'react-native-sqlite-storage';

/**
 * Opens the connection to the database
 */
const db: SQLiteDatabase = openDatabase(
  { name: 'database.db', location: 'default' },
  () => {
    Alert.alert('Conexão estabelecida', 'O banco de dados está pronto');
  },
  (error) => {
    Alert.alert(`Falha ao conectar ao banco!`, `Erro: ${error}`);
  },
);

/**
 * Executes an sql query on the database
 * @param sql SQL query to be executed
 * @param params SQL query parameters
 */
function executeQuery(sql: string, params = []): Promise<ResultSet> {
  return new Promise((resolve, reject) => {
    db.transaction((trans) => {
      trans.executeSql(
        sql,
        params,
        (trans, results) => {
          resolve(results);
        },
        (error) => {
          reject(error);
        },
      );
    });
  });
}

/**
 * Index all tables
 */
function indexTables(): Promise<ResultSet> {
  return executeQuery(
    `SELECT
    name
FROM
    sqlite_master
WHERE
    type ='table' AND
    name NOT LIKE 'sqlite_%'`,
    [],
  );
}

/**
 * Creates a new table in the database
 * @param name Table name
 */
function createTable(name: string): Promise<ResultSet> {
  return executeQuery(
    `CREATE TABLE IF NOT EXISTS ${name} (
      id integer PRIMARY KEY,
      name text NOT NULL,
      description text,
      location text,
      picture_src text,
      created_at text NOT NULL,
      updated_at text NOT NULL
  )`,
    [],
  );
}

/**
 * Drop the table if exists
 * @param name Table name
 */
function dropTable(name: string): Promise<ResultSet> {
  return executeQuery(`DROP TABLE IF EXISTS ${name}`, []);
}
export default { indexTables, createTable, dropTable };
