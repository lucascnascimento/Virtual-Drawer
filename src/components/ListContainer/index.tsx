import React from 'react';
import { View, Text } from 'react-native';
import { FlatList } from 'react-native-gesture-handler';

import { Separator } from './styles';

import ListItem from '~/components/ListItem';

import { TableName } from '~/types/types';

type ListContainerProps = { tables: Array<TableName> };

const ListContainer = ({ tables }: ListContainerProps): JSX.Element => {
  return (
    <View testID="ListContainer">
      <FlatList
        data={tables}
        keyExtractor={(table) => table.name}
        renderItem={({ item }) => <ListItem item={item} />}
        ItemSeparatorComponent={() => <Separator />}
      />
    </View>
  );
};

export default ListContainer;
