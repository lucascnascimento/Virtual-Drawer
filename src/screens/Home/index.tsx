import React, { useEffect, useState } from 'react';
import { View } from 'react-native';

import ListContainer from '~/components/ListContainer';

import db from '~/database/db';

import { TableName } from '~/types/types';

const Home = () => {
  const [tables, setTables] = useState<Array<TableName>>([]);

  useEffect(() => {
    async function getTables() {
      const res = await db.indexTables();
      const { rows } = res;
      const itemArray: TableName[] = [];
      for (let i = 1; i < rows.length; i++) {
        itemArray.push(rows.item(i));
      }
      setTables(itemArray);
    }

    getTables();
  }, []);

  return (
    <View accessibilityLabel="home screen">
      <ListContainer tables={tables} />
    </View>
  );
};

export default Home;

// todo: criar um bot√£o para adicionar tabelas e listar as tabelas
