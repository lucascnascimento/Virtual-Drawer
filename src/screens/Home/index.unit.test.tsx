/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-undef */
import React from 'react';
import { render } from 'react-native-testing-library';
import TestRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import db from '../../database/db';

import Home from './index';
import ListContainer from '~/components/ListContainer';

jest.mock('react-native-sqlite-storage');
jest.mock('../../database/db');

const initialState = {
  tables: [
    { id: 1, name: 'Primeiro Banco' },
    { id: 2, name: 'Segundo Banco' },
    { id: 3, name: 'Terceiro Banco' },
    { id: 4, name: 'Quarto Banco' },
    { id: 5, name: 'Quinto Banco' },
    { id: 6, name: 'Sexto Banco' },
    { id: 7, name: 'Sétimo Banco' },
    { id: 8, name: 'Oitavo Banco' },
    { id: 9, name: 'Nono Banco' },
    { id: 10, name: 'Décimo Banco' },
    { id: 11, name: 'Décimo Primeiro Banco' },
  ],
};

describe('render components tests', () => {
  const tree = TestRenderer.create(<Home />);
  const jsonTree = tree.toJSON();

  it('should match the snapshot', () => {
    expect(jsonTree).toMatchSnapshot();
  });

  it('should render the ListContainer component', () => {
    const testInstance = tree.root;
    expect(testInstance.findByType(ListContainer)).toBeTruthy();
  });
});

describe('database connection', () => {
  it('should load the database tables on render', () => {});
});
// it('should load 10 elements from the store', () => {
//   const db = Database.getInstance();
//   const resp = ['tabela1', 'tabela2'];
// });
