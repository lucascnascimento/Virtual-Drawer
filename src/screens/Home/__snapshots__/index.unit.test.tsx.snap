// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render components tests should match the snapshot 1`] = `
<View
  accessibilityLabel="home screen"
>
  <View
    testID="ListContainer"
  >
    <Text>
      ListContainer
    </Text>
  </View>
</View>
`;
